### PR TITLE
Make ApacheSecureAuth configurable via environment variables (LDAP)

### PR DESCRIPTION
**Pull request by h-phil**
I am trying to migrate an existing MISP deployment to the `misp-docker` containers.
Currently I use the `ApacheSecureAuth` Component and an Apache reverse proxy for LDAP authentication. (as described here: https://github.com/MISP/MISP/issues/6189)

For this to work I need the LDAP dependencies in the image and some config values in the `config.php`.

I still need to do some more testing.

### CODE REVIEW

**Comment by ostefano** [comment]
Nice!

Looking forward to it.

**Comment by h-phil** [comment]
I managed to get a working setup with my provided changes.

One thing I haven't really solved is the addition of custom CAs to the `misp-core` container. This is needed if a self-signed certificate is used by the LDAP-Server (`ldaps`).

The  workaround I used is a bit "hacky" by mounting the certs and executing `update-ca-certificates`:
```
  misp-core:
    image: ghcr.io/misp/misp-docker/misp-core:latest
    entrypoint:
      - bash
      - -c
      - "update-ca-certificates && /entrypoint.sh"
```

### Apache Reverse Proxy Setup


This is strictly speaking not part of the Misp setup, but I'll leave this here for others find:

<details>
  <summary>steps</summary>



1. Add a `httpd` container to the compose file

```yaml
  # ...

  httpd:
    image: httpd:2.4
    volumes:
      - "./httpd.conf:/usr/local/apache2/conf/httpd.conf:ro"
      - "./site.conf:/usr/local/apache2/conf/sites/site.conf:ro"
      # ldap certs
      - "./subca1.pem:/etc/ssl/certs/subca1.pem"
      - "./rootca.pem:/etc/ssl/certs/rootca.pem"
      # server certs
      - "./server-cert.pem:/etc/ssl/certs/server-cert.pem"
      - "./server-key.pem:/etc/ssl/private/server-key.pem"
      - "./ca-cert.pem:/etc/ssl/certs/ca-cert.pem"
    ports:
      - "443:443"
```

2. The `httpd.conf` is a copy of the httpd config shipped by the container with an additional line to include the sites directory

```
# standard httpd.conf shipped by the container
# ...

IncludeOptional /usr/local/apache2/conf/sites/*.conf
```

3. The `site.conf` is for a simple basic auth

```conf
LoadModule ssl_module modules/mod_ssl.so
LoadModule proxy_module modules/mod_proxy.so
LoadModule proxy_http_module modules/mod_proxy_http.so

LoadModule ldap_module modules/mod_ldap.so
LoadModule authnz_ldap_module modules/mod_authnz_ldap.so

LDAPTrustedMode SSL
LDAPVerifyServerCert On

LDAPTrustedGlobalCert CA_BASE64 "/etc/ssl/certs/subca1.pem"
LDAPTrustedGlobalCert CA_BASE64 "/etc/ssl/certs/rootca.pem"

<AuthnProviderAlias ldap auth1>
        AuthLDAPUrl "ldaps://myldapserver"
        AuthLDAPBindDN "cn=..."
        AuthLDAPBindPassword "mypassword"
</AuthnProviderAlias>

LogLevel debug
ErrorLog /dev/stderr
CustomLog /dev/stdout common

<IfModule mod_ssl.c>
        Listen 443

        <VirtualHost *:443>
                ServerName myserver.test
                ServerAdmin test@myserver.test

                SSLEngine on
                SSLCertificateFile     /etc/ssl/certs/server-cert.pem
                SSLCertificateKeyFile  /etc/ssl/private/server-key.pem
                SSLCertificateChainFile /etc/ssl/certs/ca-cert.pem

                SSLProxyEngine on
		# ONLY FOR TESTING
                SSLProxyVerify none
                SSLProxyCheckPeerName off
                SSLProxyCheckPeerCN off
                SSLProxyCheckPeerExpire off

                ProxyPreserveHost On
                RequestHeader set X-Forwarded-Proto “https”
                RequestHeader set X-Forwarded-Port “443”

                <Location />
                        ProxyPass https://misp-core:443/ nocanon
                        ProxyPassReverse https://misp-core:443/
                        SSLRequireSSL
                        AuthType Basic
                        AuthName "MISP AD authentication"
                        AuthBasicProvider auth1
                        Require valid-user
                </Location>

        </VirtualHost>
</IfModule>
```
</details>

**Comment by ostefano** [comment]
Hi @h-phil is there any place inside the entry point that can be used to run that command? Possibly inside `configure_misp.sh`?

Why the reverse proxy? To auth against LDAP twice? Would it be possible to provide apache as alternative to nginx instead?

**Comment by h-phil** [comment]
> Hi @h-phil is there any place inside the entry point that can be used to run that command? Possibly inside `configure_misp.sh`?

I have added a `update-ca-certificates` step to the `configure_misp.sh` script.

> Why the reverse proxy? To auth against LDAP twice? 

Thats how the [ApacheAuthenticate](https://github.com/MISP/MISP/blob/2.4/app/Controller/Component/Auth/ApacheAuthenticate.php) component seems to work.

From the MISP issue mentioned above:
```
The MISP LDAP/Active Directory authentication relies on two components.
 
   - Apache2 LDAP module that will perform the authentication against AD and give the username to PHP in the ENV variable REMOTE_USER
   - MISP's PHP LDAP module that will connect to AD to get more metadata on the user using the ENV setting from apache (REMOTE_USER). In short: MISP will trust Apache's user authentication decision.

As far as I understand it, currently, you cannot authenticate to LDAP with the MISP/PHP login form. (please correct me if I'm wrong)
```

> Would it be possible to provide apache as alternative to nginx instead?

I didn't really want to modify the existing `misp-core` image.

**Comment by ostefano** [comment]
CC @Kagee , appreciate a second pair of eyes on this.

**Comment by Kagee** [comment]
Looks sensible (not tested directly). Only this that may be missing is some kind of cronjob to disable users that are disabled in LDAP or has left the required groups, so their (potential) API keys stop working.

**Comment by ostefano** [comment]
@h-phil can you rebase and squash?

**Comment by h-phil** [comment]
@Kagee Good point.  There seems to be some kind of check, depending on the configuration here: https://github.com/MISP/MISP/blob/2.4/app/Controller/Component/Auth/ApacheAuthenticate.php#L173.

@ostefano I have not done this in a while. I hope that worked.

Before you merge:

One thing that I find a bit strange in this [ApacheAuthenticate](https://github.com/MISP/MISP/blob/2.4/app/Controller/Component/Auth/ApacheAuthenticate.php) is, that once you enable it, you must not allow any direct access to MISP.
Does it make sense to create a disclaimer?

**Comment by ostefano** [comment]
@h-phil the more documentation the better.

If you are doing one last pass, please also remove the empty space between the open bracket and the text on line 100 (template.env file)

**Comment by Kagee** [comment]
> once you enable it, you must not allow any direct access to MISP

~~Why?~~ You must properly unset the server env/http header that you use for auth if that is what you mean, yes.

> There seems to be some kind of check

That is only triggered when a user log in. If a user creates a auth key, loses permissions and never login again, the auth key remains valid.

> provide apache as alternative to nginx

The apachesecureauth mod does not require Apache, it is badly named. You do however need to authenticate the user and put their email into the correct server env or http header. This can be done with nginx sub auth-request hacks, or via i.e. a Apache Infront of the nginx does that does LDAP auth via basic auth or ad auth via Kerberos.

**Comment by ostefano** [comment]
Sorry for the delay here folks, I should be able to merge this before the next release of MISP 👍

**Review by ostefano — APPROVED** [review]

**Comment by ajjonen** [comment]
It doesn't actually look like the ldap_enable makes the changes to force nginx into require valid-user
```
                        AuthType Basic
                        AuthName "MISP AD authentication"
                        AuthBasicProvider auth1
                        Require valid-user
```
as this user has done with apache.

**Pull request merged**
